aten/src/ATen/MetaInit.cpp

```diff
@@ -6,23 +6,305 @@
+  Tensor surrogate = empty_like(tensor, device(kMeta));
+
+  surrogates_.emplace(ptr, surrogate);
+
+  return surrogate;
+}
+
+using TensorReplacer = std::function<Tensor(const Tensor&)>;
+
+// Replaces the tensors contained in `value` by calling `fn`.
+IValue replaceTensors(const IValue& value, const TensorReplacer& fn) {
+  bool replaced = false;
+
+  IValue::HashAliasedIValueMap memo{};
+
+  auto visitor = [&replaced, &memo, &fn](const IValue& v) {
+    // Deep-copy all compound objects.
+    if (v.isTuple() || v.isList() || v.isGenericDict()) {
+      return false;
+    }
+
+    if (v.isTensor()) {
+      const Tensor& inp = v.toTensor();
+
+      Tensor out = fn(inp);
+      // If not replaced, we need to put the original value into `memo` so that
+      // it does not get cloned by `deepcopy()`.
+      if (out.is_same(inp)) {
+        memo[v] = v;
+      } else {
+        memo[v] = out;
+
+        replaced = true;
+      }
+    } else {
+      // Do not clone other non-compound objects.
+      memo[v] = v;
+    }
+
+    return false;
+  };
+
+  value.visit(visitor);
+
+  if (replaced) {
+    return value.deepcopy(memo);
+  } else {
+    return value;
+  }
+}
+
+void replaceTensorArguments(const OperatorHandle& op, Stack& s, const TensorReplacer& fn) {
+  std::size_t num_arguments = op.schema().arguments().size();
+
+  // Perform an in-place update of the tensor arguments in the stack.
+  for (std::size_t i = 0; i < num_arguments; i++) {
+    IValue& arg = torch::jit::peek(s, i, num_arguments);
+
+    arg = replaceTensors(arg, fn);
+  }
+}
+
+// Replaces all foreign tensor arguments with surrogate tensors.
+void replaceForeignTensorArguments(const OperatorHandle& op, Stack& s) {
+  auto replace_foreign_tensor = [](const Tensor& tensor) -> Tensor {
+    // If the tensor was not output by an operation that we recorded, we should
+    // replace it with a surrogate tensor.
+    if (tensor.is_meta()) {
+      return tensor;
+    } else {
+      return meta_init_state.getOrCreateSurrogateTensor(tensor);
+    }
+  };
+
+  replaceTensorArguments(op, s, replace_foreign_tensor);
+}
+
+// Checks whether a `TensorOptions` argument can be parsed from the individual
+// arguments of `schema`.
+bool hasTensorOptions(const FunctionSchema& schema) noexcept {
+  std::array<c10::string_view, 4> tensor_opts = {{
+      "dtype", "layout", "device", "pin_memory"
+  }};
+
+  const std::vector<Argument>& args = schema.arguments();
+  if (args.size() < tensor_opts.size()) {
+    return false;
+  }
+
+  // Checks if the arguments starting at `arg_pos` represent a `TensorOptions`.
+  auto are_tensor_opts = [&tensor_opts](auto arg_pos) noexcept {
+    for (const auto& tensor_opt : tensor_opts) {
+      if (tensor_opt != arg_pos->name()) {
+        return false;
+      }
+      ++arg_pos;
+    }
+    return true;
+  };
+
+  for (auto pos = args.begin(); pos <= args.end() - tensor_opts.size(); ++pos) {
```

Comment:
Are we relying on a certain order of arguments here that we exclude the last few arguments from this logic?

Reply:
Yes, I am using the same logic as in [`tools/codegen/models.py`](https://github.com/pytorch/pytorch/blob/master/tools/codegen/model.py#L1401) to determine a `TensorOptions` parameter from the raw Python parameters. Then this is used in the same way as in [`tools/autograd/gen_variable_factories.py`](https://github.com/pytorch/pytorch/blob/master/tools/autograd/gen_variable_factories.py#L48) to find out whether an operator is a tensor factory. In summary I use the same logic as Autograd's C++ function generator to determine tensor factories - albeit during runtime instead of compilation time. Just to be even more sure I also check whether the operator returns at least one tensor.

Reply:
As discussed before, although I am fairly certain that this logic will work almost always, once we have operator tagging in place, ideally this implementation should be simplified to use it.